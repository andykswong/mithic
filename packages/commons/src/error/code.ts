/**
 * Common error codes.
 * @packageDocumentation
 */

/** Generic error code. */
export const ERR = 'ERR';

/** Abort error code. */
export const ABORT_ERR = 'ABORT_ERR';

/** Timeout error code. */
export const TIMEOUT_ERR = 'TIMEOUT_ERR';

/** Not found error code. */
export const NOT_FOUND_ERR = 'NOT_FOUND_ERR';

/** Not supported error code. */
export const NOT_SUPPORTED_ERR = 'NOT_SUPPORTED_ERR';

/** Invalid state error code. */
export const INVALID_STATE_ERR = 'INVALID_STATE_ERR';

/** Generic operation failed error code. */
export const ERR_OPERATION_FAILED = 'ERR_OPERATION_FAILED';

/** Error code for missing dependency. */
export const ERR_DEPENDENCY_MISSING = 'ERR_DEPENDENCY_MISSING';

/** Error code for already exist content. */
export const ERR_EXIST = 'ERR_EXIST';
