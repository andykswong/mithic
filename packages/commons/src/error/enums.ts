/** Error names. */
export enum ErrorName {
  Coded = 'CodedError',
  Abort = 'AbortError',
  InvalidState = 'InvalidStateError',
  NotFound = 'NotFoundError',
  OpFailed = 'OperationError',
}

/** Error codes. */
export enum ErrorCode {
  Error = 'ERR',
  Abort = 'ABORT_ERR',
  CryptoInvalidIV = 'ERR_CRYPTO_INVALID_IV',
  CryptoInvalidKey = 'ERR_CRYPTO_INVALID_KEYPAIR',
  CryptoKeyLen = 'ERR_CRYPTO_INVALID_KEYLEN',
  CryptoMsgLen = 'ERR_CRYPTO_INVALID_MESSAGELEN',
  Exist = 'ERR_EXIST',
  InvalidArg = 'ERR_INVALID_ARG_VALUE',
  InvalidState = 'ERR_INVALID_STATE',
  NotFound = 'ERR_NOT_FOUND',
  OpFailed = 'ERR_OPERATION_FAILED',
  UnsupportedOp = 'ERR_UNSUPPORTED_OPERATION',
}
