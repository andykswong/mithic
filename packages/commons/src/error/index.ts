export * from './enums.js';
export * from './error.js';
export * from './errors.js';
