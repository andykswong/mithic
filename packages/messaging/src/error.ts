import { CodedError } from '@mithic/commons';

/** Message validation error. */
export class MessageValidationError<T = unknown, E = unknown> extends CodedError<T, E> {
  constructor(
    message?: string,
    options?: MessageValidationErrorOptions<T>,
  ) {
    super(message, {
      name: MessageValidationError.name,
      ...options,
      code: options?.code ?? MessageValidationErrorCode.Reject,
    });
  }
}

/** Options for initializing a {@link MessageValidationError}. */
export interface MessageValidationErrorOptions<T> extends ErrorOptions {
  /** Error code. */
  code?: MessageValidationErrorCode;

  /** Error details. */
  detail?: T;
}

/** Message validation error code. */
export enum MessageValidationErrorCode {
  /** The message should be ignored. */
  Ignore = 'ignore',

  /** The message is considered invalid, and it should be rejected. */
  Reject = 'reject'
}
