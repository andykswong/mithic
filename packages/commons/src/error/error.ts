import { ERR } from './code.ts';

/** Error with error code. */
export class CodedError<T = unknown, E = unknown> extends Error {
  /** Thr error code. */
  public readonly code: string;

  /** The error cause. */
  public override readonly cause?: E;

  /** Returns any data passed when initializing the error. */
  public detail?: T;

  public constructor(
    message?: string,
    options?: CodedErrorOptions<T>,
  ) {
    super(message, options);
    this.name = options?.name ?? CodedError.name;
    this.code = options?.code ?? ERR;
    this.detail = options?.detail;
  }
}

/** Options for initializing a {@link CodedError}. */
export interface CodedErrorOptions<T> extends ErrorCodeDetailOptions<T> {
  /** Error name. */
  readonly name?: string;
}

/** Options for initializing an error with code and detail. */
export interface ErrorCodeDetailOptions<T> extends ErrorDetailOptions<T> {
  /** Error code. */
  readonly code?: string;
}

/** Options for initializing an error with detail. */
export interface ErrorDetailOptions<T> extends ErrorOptions {
  /** Error details. */
  readonly detail?: T;
}
