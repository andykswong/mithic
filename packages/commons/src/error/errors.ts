import { ErrorCode, ErrorName } from './enums.js';
import { CodedError } from './error.js';

/** Creates an AbortError. */
export function abortError<T = unknown, E = unknown>(reason: string = ErrorName.Abort, cause?: E): CodedError<T, E> {
  return new CodedError(reason, {
    name: ErrorName.Abort,
    code: ErrorCode.Abort,
    cause,
  });
}

/** Creates a NotFoundError. */
export function notFoundError<T, E>(
  reason: string = ErrorName.NotFound, detail?: T, cause?: E
): CodedError<T, E> {
  return new CodedError(reason, {
    name: ErrorName.NotFound,
    code: ErrorCode.NotFound,
    detail,
    cause,
  });
}

/** Creates an InvalidStateError. */
export function invalidStateError<T, E>(
  reason: string = ErrorName.InvalidState, code: string = ErrorCode.InvalidState, detail?: T, cause?: E
): CodedError<T, E> {
  return new CodedError(reason, {
    name: ErrorName.InvalidState,
    code,
    detail,
    cause,
  });
}

/** Creates an OperationError. */
export function operationError<T, E>(
  reason: string = ErrorName.OpFailed, code: string = ErrorCode.OpFailed, detail?: T, cause?: E
): CodedError<T, E> {
  return new CodedError(reason, {
    name: ErrorName.OpFailed,
    code,
    detail,
    cause,
  });
}
