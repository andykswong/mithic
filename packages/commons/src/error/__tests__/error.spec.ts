import { describe, expect, it } from '@jest/globals';
import { ErrorCode } from '../enums.js';
import { CodedError } from '../error.js';

describe(CodedError.name, () => {
  it('should initialize with given options', () => {
    const cause = new Error('Error cause');
    const error = new CodedError('Error message', {
      name: 'TestError',
      code: ErrorCode.Abort,
      detail: { 'this': 'is a testing' },
      cause
    });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});
