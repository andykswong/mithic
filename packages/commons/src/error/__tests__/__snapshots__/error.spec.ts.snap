// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodedError should initialize with given options 1`] = `[TestError: Error message]`;

exports[`CodedError should initialize with given options 2`] = `
{
  "code": "ABORT_ERR",
  "detail": {
    "this": "is a testing",
  },
  "name": "TestError",
}
`;
